### PR TITLE
swaps: client fsm cleanup and durable sessions

### PR DESCRIPTION
**Pull request by bhandras**
## Summary

This PR adds the client-side swap SDK that pairs with the server-side swap
state machines in `swapdk-server`.

The package lives under `sdk/swaps` and exposes two high-level flows:

- `PayViaLightning`: client pays a Lightning invoice by funding an Ark vHTLC.
  This is the client counterpart to the server **in-swap** flow.
- `ReceiveViaLightning`: client receives over Lightning by claiming a
  server-funded Ark vHTLC. This is the client counterpart to the server
  **out-swap** flow.

The branch also adds the small supporting pieces needed by those flows:

- shared `swaprpc` protobuf bindings for the client/server protocol
- deterministic Ark tree-path conversion helpers
- OOR snapshot/transaction-context fixes for custom vHTLC spends
- an isolated SQL-backed swap session store under `sdk/swaps`
- CLI commands that start, list, and resume persisted swap sessions without duplicating lifecycle logic

## Review Map

The stack is intentionally split so review can proceed from low-level helpers
up to the state machines:

1. Build and dependency updates for Loop FSM, lndclient, gRPC, and protobuf.
2. `swaprpc` protocol definitions and generated bindings.
3. Pure Ark tree conversion helpers.
4. OOR durability for custom spend sequence/locktime requirements.
5. OOR transaction-context handling for custom spends.
6. A small `sdk/ark` height helper used by timeout checks.
7. `sdk/swaps` public API, server transport, invoice abstraction, and helpers.
8. The isolated SQLite/sqlc session store.
9. Pay-side FSM descriptor.
10. Pay-side lifecycle implementation.
11. Pay-side unit coverage.
12. Receive-side FSM descriptor.
13. Receive-side lifecycle implementation.
14. Receive-side unit coverage.
15. CLI wiring for starting, listing, and resuming swaps.
16. Generated schema comment refresh.

The first eight commits establish the substrate. The next six are the main
swap review: FSM graph first, execution second, tests third for each direction.

## Architecture

`sdk/swaps` is deliberately an orchestration layer. It does not embed Ark
wallet logic and it does not import `swapdk-server` internals.

- Swap-server calls go through the generated `swaprpc` client wrapped by
  `SwapServerConn`.
- Ark wallet/indexer calls go through `sdk/ark` via the `DaemonConn` boundary.
- Lightning invoice creation is injected through `InvoiceCreator`, so the swap
  package does not own a specific LN wallet implementation.
- Durable client state is kept in a separate SQLite database owned by
  `sdk/swaps`, not in the Ark daemon database.
- List APIs expose pay and receive progress, fees, outpoints, and terminal
  context without requiring callers to inspect sqlc rows.

This lets the swap client run in the same binary as an embedded Ark client or
against a remote Ark daemon. Higher-level applications can choose the hosting
model without changing the swap FSMs.

## Durability Model

Both directions follow the same restart rule:

> Persist intent before an external side effect, then reconcile daemon/indexer
> state before retrying after restart.

Examples:

- Funding intent is recorded before submitting the OOR transfer that funds a
  pay-side vHTLC.
- Claim intent is recorded before spending a receive-side vHTLC with the
  invoice preimage.
- Refund intent is recorded before spending an expired pay-side vHTLC through
  the timeout refund path.
- Terminal sessions short-circuit resume so completed, expired, refunded,
  failed, and intervention sessions do not resurrect.

The store persists the negotiated vHTLC config, policy template, script,
outpoint, amounts, preimage/claim/refund metadata, timestamps, and intervention
reason. That gives resume enough context to adopt already-submitted work instead
of replaying it blindly.

## PayViaLightning FSM

`PayViaLightning` is the client-side Ark → Lightning flow. The client asks the
server to create an in-swap, funds the returned vHTLC policy output, waits for
the server to pay the Lightning invoice and claim with the preimage, and refunds
its vHTLC if the swap times out before the server claims.

```mermaid
stateDiagram-v2
    [*] --> Created

    Created --> Created: OnAdvance
    Created --> SwapCreated: OnSwapCreated
    Created --> Expired: OnExpired
    Created --> Failed: OnFailed

    SwapCreated --> SwapCreated: OnAdvance
    SwapCreated --> FundingInitiated: OnFundingInitiated
    SwapCreated --> VHTLCFunded: OnVHTLCFunded
    SwapCreated --> Completed: OnCompleted
    SwapCreated --> Expired: OnExpired
    SwapCreated --> RefundInitiated: OnRefundInitiated
    SwapCreated --> NeedsIntervention: OnNeedsIntervention
    SwapCreated --> Failed: OnFailed

    FundingInitiated --> FundingInitiated: OnAdvance
    FundingInitiated --> VHTLCFunded: OnVHTLCFunded
    FundingInitiated --> Completed: OnCompleted
    FundingInitiated --> Expired: OnExpired
    FundingInitiated --> RefundInitiated: OnRefundInitiated
    FundingInitiated --> NeedsIntervention: OnNeedsIntervention
    FundingInitiated --> Failed: OnFailed

    VHTLCFunded --> VHTLCFunded: OnAdvance
    VHTLCFunded --> WaitingForClaim: OnWaitForClaim
    VHTLCFunded --> Completed: OnCompleted
    VHTLCFunded --> Expired: OnExpired
    VHTLCFunded --> RefundInitiated: OnRefundInitiated
    VHTLCFunded --> NeedsIntervention: OnNeedsIntervention
    VHTLCFunded --> Failed: OnFailed

    WaitingForClaim --> WaitingForClaim: OnAdvance
    WaitingForClaim --> Completed: OnCompleted
    WaitingForClaim --> Expired: OnExpired
    WaitingForClaim --> RefundInitiated: OnRefundInitiated
    WaitingForClaim --> Refunded: OnRefunded
    WaitingForClaim --> NeedsIntervention: OnNeedsIntervention
    WaitingForClaim --> Failed: OnFailed

    RefundInitiated --> RefundInitiated: OnAdvance
    RefundInitiated --> Completed: OnCompleted
    RefundInitiated --> Refunded: OnRefunded
    RefundInitiated --> NeedsIntervention: OnNeedsIntervention
    RefundInitiated --> Failed: OnFailed

    Completed --> [*]
    Expired --> [*]
    Refunded --> [*]
    NeedsIntervention --> [*]
    Failed --> [*]
```

### Pay Timeout Semantics

Pay-side timeout is recoverable when the client funded the vHTLC but the server
did not claim it in time. The client moves through `RefundInitiated` and then
`Refunded` after the Ark refund spend is submitted or observed.

`NeedsIntervention` is reserved for anomalous data that should not be treated as
an ordinary timeout, such as an unexpected spender or inconsistent indexed data.

## ReceiveViaLightning FSM

`ReceiveViaLightning` is the client-side Lightning → Ark flow. The client
prepares a preimage-backed invoice and expected vHTLC script, waits for the
server to fund that vHTLC, then claims the Ark output with the invoice preimage.

```mermaid
stateDiagram-v2
    [*] --> Created

    Created --> Created: OnAdvance
    Created --> InvoiceCreated: OnInvoiceCreated
    Created --> Expired: OnExpired
    Created --> NeedsIntervention: OnNeedsIntervention
    Created --> Failed: OnFailed

    InvoiceCreated --> InvoiceCreated: OnAdvance
    InvoiceCreated --> VHTLCFunded: OnVHTLCFunded
    InvoiceCreated --> Expired: OnExpired
    InvoiceCreated --> NeedsIntervention: OnNeedsIntervention
    InvoiceCreated --> Failed: OnFailed

    VHTLCFunded --> VHTLCFunded: OnAdvance
    VHTLCFunded --> ClaimInitiated: OnClaimInitiated
    VHTLCFunded --> Expired: OnExpired
    VHTLCFunded --> NeedsIntervention: OnNeedsIntervention
    VHTLCFunded --> Failed: OnFailed

    ClaimInitiated --> ClaimInitiated: OnAdvance
    ClaimInitiated --> Completed: OnCompleted
    ClaimInitiated --> Expired: OnExpired
    ClaimInitiated --> NeedsIntervention: OnNeedsIntervention
    ClaimInitiated --> Failed: OnFailed

    Completed --> [*]
    Expired --> [*]
    NeedsIntervention --> [*]
    Failed --> [*]
```

### Receive Timeout Semantics

Before funding, receive expiry simply means the route/invoice lifetime elapsed
and the flow can terminalize as `Expired`.

After funding, the client either claims with the preimage or records enough
context for a safe terminal state. Unknown or inconsistent spend data is kept as
`NeedsIntervention` so the caller does not lose the outpoint and reason needed
for operator review.

## Test Coverage

The unit tests cover both state machines locally without requiring the server
itest harness for every edge case.

Pay-side coverage includes:

- happy path completion and preimage validation
- resume from persisted sessions
- terminal-state resume short-circuiting
- interrupted funding adoption
- funding deadline and refund-locktime checks
- timeout refund through `RefundInitiated` → `Refunded`
- intervention reasons for unsafe or inconsistent indexed data

Receive-side coverage includes:

- invoice preparation and vHTLC script derivation
- funded-vHTLC observation and amount validation
- claim intent persistence before the Ark spend
- claim retry/adoption after restart
- terminal-state resume short-circuiting
- expiry and intervention paths

The paired end-to-end coverage lives in the `swapdk-server` PRs:

- `lightninglabs/swapdk-server#9`: server out-swap / client receive flow
- `lightninglabs/swapdk-server#12`: server in-swap / client pay flow

## Verification

Latest local verification after reshaping the stack:

```bash
make commitmsg-lint range=origin/main..HEAD
make lint
go test ./sdk/swaps ./arkrpc/treeconv ./oor -count=1 -timeout=120s
go test ./cmd/darepocli/darepoclicommands -count=1 -timeout=120s
git diff --check origin/main...HEAD
```

The reshaped branch preserves the exact same final tree as the previous pushed
head; only the commit story changed.

### CODE REVIEW

File: sdk/swaps/in_swap.go [hunk 1]
```diff
@@ -0,0 +1,772 @@
...
+	if err != nil {
+		return fmt.Errorf("build vHTLC policy: %w", err)
+	}
+
+	pkScript, err := policy.PkScript()
+	if err != nil {
+		return fmt.Errorf("get vHTLC pkScript: %w", err)
+	}
+
+	policyTemplate, err := encodeVHTLCPolicyTemplate(policy)
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sdk/swaps/in_swap.go:551]: The remaining encode here is the initial template derivation for the pay session. The later funding path now uses the persisted session template instead of re-encoding it, so there is no second in-memory encoding path to drift before funding.

File: sdk/swaps/out_swap.go [hunk 2]
```diff
@@ -0,0 +1,724 @@
...
+		slog.String("outpoint", outpoint),
+		slog.Int64("amount", amount),
+	)
+
+	claimPath, err := policy.ClaimPath(preimage)
+	if err != nil {
+		return fmt.Errorf("build claim path: %w", err)
+	}
+
+	policyTemplate, err := encodeVHTLCPolicyTemplate(policy)
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sdk/swaps/out_swap.go]: Fixed in the folded update: the receive-side claim path now reuses the template derived during invoice preparation instead of re-encoding the policy at claim time.

File: sdk/swaps/daemon_conn.go [hunk 3]
```diff
@@ -0,0 +1,391 @@
...
+// parseSessionTxid converts a human-readable txid string into the raw
+// chainhash byte order expected by the daemon/indexer RPC boundary.
+func parseSessionTxid(sessionTxid string) ([]byte, error) {
+	hash, err := chainhash.NewHashFromStr(sessionTxid)
+	if err != nil {
+		return nil, err
+	}
+
+	return append([]byte(nil), hash[:]...), nil
+}
```
[the remark this answers is not in the document]
Comment (bhandras) [reply sdk/swaps/daemon_conn.go]: This became obsolete after the stack was consolidated onto the sdk/ark facade: the old daemon_conn helper and parseSessionTxid path are no longer part of sdk/swaps.

**Comment by bhandras** [comment]
@claude pls re-review

**Comment by bhandras** [comment]
Thanks Claude, fixed and folded the actionable items from this pass.

- Removed the `"indexed-claim"` pseudo session id path. Already-indexed receive claims now return an explicit internal signal and transition to `Completed` without persisting a synthetic OOR session id.
- Receive sessions now persist and restore `vhtlc_policy_template`, matching pay-session behavior and avoiding policy-template re-derivation on resume.
- Documented the ignored invoice-generator compatibility parameters, clarified the `NewSwapClient` invoice-creator requirement, and added the CLI comment explaining why the ephemeral invoice signer is safe for resumed sessions.
- Promoted the pre-funding `observeLiveVHTLC` query failure from debug to warn.
- Added a defensive comment around the current OOR custom spend assumption that leaves use either CSV or CLTV tx context, not both.

I intentionally kept the pay-side amount-mismatch path as refund-driven rather than `NeedsIntervention`: at that point the client has not received a claim preimage, the mismatch is economically recoverable, and our current client/server timeout design is to auto-sweep back whenever no party has irreversibly lost funds. The refund path still preserves the reason/outpoint for audit.

Validation:

- `go test ./sdk/swaps ./oor ./cmd/darepocli/darepoclicommands -count=1 -timeout=120s`
- `make lint`
- `make commitmsg-lint range=origin/main..HEAD`

**Comment by bhandras** [comment]
@claude re-review

**Comment by bhandras** [comment]
Folded the latest review feedback into the existing 17-commit stack and force-pushed the rewritten branch.

What changed:

- Replaced `arkrpc/treeconv`'s duplicate tree-path conversion code with a compatibility shim over the canonical `arkrpc` converter.
- Fixed `mutateAndPersist` rollback semantics so both pay and receive sessions restore their in-memory snapshot when the mutation closure fails.
- Shared custom-spend sequence selection between checkpoint txs and Ark txs, including the CLTV non-final-sequence fallback.
- Added `ErrSwapRefunded` so pay-side refunds no longer surface as generic swap expiry.
- Tightened invoice/receive validation for amounts and route-hint integer widths before BOLT-11 construction.
- Documented the receive invoice preparation ordering exception: route hints are lease-like and invoice generation is stateless, with persistence before returning the invoice to the caller.
- Simplified receive claim funding remembrance into one merged persistence call.
- Documented the dual meaning of `refundSessionID` / `RefundSessionID` when a refund is adopted from the indexer.
- Derived Loop FSM transition descriptors from the durable transition maps for pay and receive flows.
- Unified direct and signer-backed invoice creation behind shared BOLT-11 assembly.
- Made the CLI use TLS for remote swap-server connections by default, while keeping local development plaintext and adding `--swapserver-tlscert` / `--swapserver-insecure` flags.
- Added early vHTLC config sanity checks and documented session single-goroutine usage.

Local verification passed:

- `make commitmsg-lint range=origin/main..HEAD`
- `go test ./arkrpc/treeconv ./sdk/swaps ./oor ./cmd/darepocli/darepoclicommands -count=1`
- `make lint-local`
- `make unit`

GitHub CI is green on the pushed stack; `claude-review` is skipped as expected.

@claude pls re-review

**Comment by bhandras** [comment]
@claude pls re-review
